File: captainhcg/go/22-generate-parentheses.go
var res []string

func generateParenthesis(n int) []string {
    res = make([]string, 0)
    genHelper(n, n, "")
    return res
}

func genHelper(left, right int, cur string){
    if left == 0 && right == 0 {
        res = append(res, cur)
        return
    }
    if left == 0 {
        genHelper(left, right - 1, cur + ")")
    } else if left < right {
        genHelper(left, right - 1, cur + ")")
        genHelper(left - 1, right, cur + "(")
    } else if left == right {
        genHelper(left - 1, right, cur + "(")
    }
}

File: bwu/dfs/22-generate-parentheses.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        res = set()
        self.dfs(['('] * n, [')'] * n, '', res)
        return list(res)
        
    def dfs(self, left, right, tmp, res):
        if not len(left) and not len(right) and tmp not in res:
            res.add(tmp)
        elif not len(left) and len(right):
            self.dfs([], right[: -1], tmp + ')', res)
        elif len(left) <= len(right):
            self.dfs(left[: -1], right, tmp + '(', res)
            self.dfs(left, right[: -1], tmp + ')', res)
